Plutarch/Verify.hs

```diff
@@ -0,0 +1,277 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-unticked-promoted-constructors #-}
+{-# OPTIONS_GHC -Wredundant-constraints #-}
+
+module Plutarch.Verify (
+  PTryFrom (PExcess, ptryFrom),
+  pcheckType,
+  pcheckByteStr,
+  pcheckInt,
+  pcheckList,
+  pcheckMap,
+) where
+
+import Plutarch.Builtin (
+  PAsData,
+  PBuiltinList,
+  PBuiltinMap,
+  PBuiltinPair,
+  PData,
+  PIsData (pfromData),
+  pasByteStr,
+  pasInt,
+  pasList,
+  pasMap,
+  pdata,
+  pforgetData,
+  pfstBuiltin,
+  ppairDataBuiltin,
+  psndBuiltin,
+ )
+
+import Plutarch.ByteString (PByteString)
+import Plutarch.Integer (PInteger)
+import Plutarch.Internal.Other (
+  POpaque,
+  PType,
+  Term,
+  pcon,
+  perror,
+  pforce,
+  phoistAcyclic,
+  plam,
+  plet,
+  (#),
+  type (:-->),
+ )
+import Plutarch.Unit (PUnit (PUnit))
+
+import Plutarch.DataRepr.Internal (
+  PDataRecord,
+  PLabeledType ((:=)),
+  pdcons,
+  pdnil,
+ )
+
+import Plutarch.Bool (pif, (#==))
+
+import Plutarch.List (
+  phead,
+  pmap,
+  ptail,
+ )
+
+import Plutarch.Unsafe (punsafeBuiltin, punsafeCoerce)
+import qualified PlutusCore as PLC
+
+import Plutarch.TermCont (TermCont, tcont, unTermCont)
+
+{- |
+    Note: PAsData POpaque ~ PData
+-}
+
+----------------------- The class PTryFrom ----------------------------------------------
+
+{- |
+    This checks the datastructure for validity.
+    If you don't care about parts of the datastructure
+    don't verify those parts, just let it return a PData
+    instead
+    Be aware this might get really expensive, so only
+    use it if you cannot establish trust otherwise
+    (e.g. via only checking a part of your Data with
+    PTryFrom)
+-}
+class PTryFrom (a :: PType) (b :: PType) where
+  --     "safest" type --^            ^-- target type
+  type PExcess a b :: PType
+  ptryFrom :: Term s a -> TermCont s (Term s b, Term s (PExcess a b))
+
+----------------------- PData instances -------------------------------------------------
+
+instance PTryFrom PData (PAsData PInteger) where
+  type PExcess PData (PAsData PInteger) = PInteger
+  ptryFrom opq = do
+    ver <- tcont $ plet (pasInt # opq)
+    pure $ (punsafeCoerce opq, ver)
+
+instance PTryFrom PData (PAsData PByteString) where
+  type PExcess PData (PAsData PByteString) = PByteString
+  ptryFrom opq = do
+    ver <- tcont $ plet (pasByteStr # opq)
+    pure $ (punsafeCoerce opq, ver)
+
+instance PTryFrom PData PData where
+  type PExcess PData PData = PUnit
+  ptryFrom opq = pure $ (opq, pcon PUnit)
```

Comment:
this feels strange

Reply:
you could do `PTryFrom a a`. Not sure we need this instance thoughh.

Reply:
I can just remove it but I don't see the issue with it, maybe we'll need it as a base case for a structure that doesn't rely on the inner type being wrapped in `PAsData`? (Although I can't come up with a usecase, either :D )